// react-responsive-ui/commonjs/FileUpload.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var File_upload = function (_PureComponent) {
	(0, _inherits3.default)(File_upload, _PureComponent);

	function File_upload() {
		(0, _classCallCheck3.default)(this, File_upload);

		var _this = (0, _possibleConstructorReturn3.default)(this, (File_upload.__proto__ || (0, _getPrototypeOf2.default)(File_upload)).call(this));

		_this.on_file_selected = _this.on_file_selected.bind(_this);
		_this.on_click = _this.on_click.bind(_this);
		_this.on_input_click = _this.on_input_click.bind(_this);
		_this.on_input_change = _this.on_input_change.bind(_this);
		return _this;
	}

	(0, _createClass3.default)(File_upload, [{
		key: 'on_input_click',
		value: function on_input_click(event) {
			event.stopPropagation();
		}
	}, {
		key: 'on_input_change',
		value: function on_input_change(event) {
			this.on_file_selected(event);
		}
	}, {
		key: 'on_file_selected',
		value: function on_file_selected(event) {
			var _props = this.props,
			    action = _props.action,
			    multiple = _props.multiple;


			var data = event.target.files;

			if (!multiple) {
				data = data[0];
			}

			action(data);

			// reset the selected file
			// so that onChange would trigger again
			// even with the same file
			event.target.value = null;
		}
	}, {
		key: 'on_click',
		value: function on_click(event) {
			var _props2 = this.props,
			    disabled = _props2.disabled,
			    onClick = _props2.onClick;


			if (disabled) {
				return event.preventDefault();
			}

			if (onClick) {
				onClick();
			}

			this.file_upload.click();
		}
	}, {
		key: 'render',
		value: function render() {
			var _this2 = this;

			var _props3 = this.props,
			    disabled = _props3.disabled,
			    style = _props3.style,
			    className = _props3.className,
			    children = _props3.children;


			var markup = _react2.default.createElement(
				'div',
				{
					style: style,
					className: (0, _classnames2.default)('rrui__file-upload', className, {
						'rrui__file-upload--disabled': disabled
					}),
					onClick: this.on_click },
				_react2.default.createElement('input', {
					type: 'file',
					ref: function ref(_ref) {
						return _this2.file_upload = _ref;
					},
					key: 'file_input',
					style: input_style,
					onClick: this.on_click,
					onChange: this.on_input_change }),
				children
			);

			return markup;
		}
	}]);
	return File_upload;
}(_react.PureComponent);

File_upload.propTypes = {
	// On file chosen handler
	action: _propTypes2.default.func.isRequired,

	// Allows choosing multiple files if `true`
	multiple: _propTypes2.default.bool,

	// Disables the file input
	disabled: _propTypes2.default.bool,

	// `onClick` handler
	onClick: _propTypes2.default.func,

	// The clickable area, like "Click here to choose a file"
	children: _propTypes2.default.node,

	// CSS class
	className: _propTypes2.default.string,

	// CSS style object
	style: _propTypes2.default.object
};
exports.default = File_upload;


var input_style = {
	display: 'none'
};
