// react-responsive-ui/commonjs/Form.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Prevents `<form/> submission when `busy` is `true`.
// And also inserts `<Form.Error/>` when `error` is passed.
// Using `Component` here instead of `PureComponent`
// because `<Form.Actions>` depends on `context` and therefore
// should be rerendered even if the `props` haven't changed.
// And probably that was also why it would keep displaying an old error
// (and not resetting it for some reason).
var Form = function (_Component) {
	(0, _inherits3.default)(Form, _Component);

	function Form() {
		(0, _classCallCheck3.default)(this, Form);

		var _this = (0, _possibleConstructorReturn3.default)(this, (Form.__proto__ || (0, _getPrototypeOf2.default)(Form)).call(this));

		_this.submit = _this.submit.bind(_this);
		_this.on_key_down = _this.on_key_down.bind(_this);
		return _this;
	}

	(0, _createClass3.default)(Form, [{
		key: 'render',
		value: function render() {
			var _props = this.props,
			    post = _props.post,
			    error = _props.error,
			    className = _props.className,
			    style = _props.style;


			var markup = _react2.default.createElement(
				'form',
				{
					onSubmit: this.submit,
					onKeyDown: this.on_key_down,
					action: post,
					method: 'POST',
					className: (0, _classnames2.default)('rrui__form', className),
					style: style,
					noValidate: true },
				this.render_children(error)
			);

			return markup;
		}

		// Adds `error` element to the form

	}, {
		key: 'render_children',
		value: function render_children(error) {
			var children = this.props.children;


			var form_elements = _react2.default.Children.toArray(children);

			// Insert `error` element
			if (error) {
				// Will be set to `null` upon insertion
				var error_element = _react2.default.createElement(
					Form.Error,
					{ key: 'form-error' },
					error
				);

				// Show form error above form actions,
				// so that the error will be visible and won't be overlooked.
				var index = 0;
				var _iteratorNormalCompletion = true;
				var _didIteratorError = false;
				var _iteratorError = undefined;

				try {
					for (var _iterator = (0, _getIterator3.default)(form_elements), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
						var child = _step.value;

						if (child.type === Form.Error) {
							form_elements[index] = _react2.default.cloneElement(child, { key: 'form-error' }, error);
							error_element = null;
							break;
						}

						if (child.type === Form.Actions) {
							form_elements.insert_at(index, error_element);
							error_element = null;
							break;
						}

						index++;
					}
				} catch (err) {
					_didIteratorError = true;
					_iteratorError = err;
				} finally {
					try {
						if (!_iteratorNormalCompletion && _iterator.return) {
							_iterator.return();
						}
					} finally {
						if (_didIteratorError) {
							throw _iteratorError;
						}
					}
				}

				if (error_element) {
					form_elements.push(error_element);
				}
			}

			return form_elements;
		}
	}, {
		key: 'submit',
		value: function submit(event) {
			if (event) {
				event.preventDefault();
			}

			var _props2 = this.props,
			    busy = _props2.busy,
			    submit = _props2.submit;

			// Prevent form double submit
			// (because not only buttons submit a form,
			//  therefore just disabling buttons is not enough).

			if (busy) {
				return;
			}

			if (submit) {
				return submit();
			}
		}
	}, {
		key: 'on_key_down',
		value: function on_key_down(event) {
			// Cancel editing on "Escape" key
			if (event.keyCode === 27) {
				var cancel = this.props.cancel;


				if (cancel) {
					cancel();
				}

				event.preventDefault();
			}
		}
	}]);
	return Form;
}(_react.Component);

Form.propTypes = {
	// `onSubmit` handler
	submit: _propTypes2.default.func,

	// On `Escape` keydown handler
	cancel: _propTypes2.default.func,

	// `error` can be passed for non-javascript old-school forms.
	// e.g. when a form submitted via HTTP POST request had an error,
	// then this error is rendered as part of the form.
	error: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.bool, _propTypes2.default.node]),

	// The HTML form `action` attribute,
	// i.e. the URL to which the form will be posted
	// in case of a non-javascript old-school POST submission.
	post: _propTypes2.default.string,

	// Is form submission in progress
	busy: _propTypes2.default.bool.isRequired,

	// CSS class
	className: _propTypes2.default.string,

	// CSS style object
	style: _propTypes2.default.object
};
Form.defaultProps = {
	busy: false
};
exports.default = Form;


Form.Error = function (_ref) {
	var children = _ref.children;

	return _react2.default.createElement(
		'div',
		{ className: 'rrui__form__error' },
		children
	);
};

// Using `Component` here instead of `PureComponent`
// because it depends on `context` and therefore should be
// rerendered even if the `props` haven't changed.
Form.Actions = function (_Component2) {
	(0, _inherits3.default)(Actions, _Component2);

	function Actions() {
		(0, _classCallCheck3.default)(this, Actions);
		return (0, _possibleConstructorReturn3.default)(this, (Actions.__proto__ || (0, _getPrototypeOf2.default)(Actions)).apply(this, arguments));
	}

	(0, _createClass3.default)(Actions, [{
		key: 'componentWillMount',
		value: function componentWillMount() {
			var rrui__modal = this.context.rrui__modal;


			if (rrui__modal) {
				rrui__modal.register_form();
			}
		}
	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			var rrui__modal = this.context.rrui__modal;


			if (rrui__modal) {
				rrui__modal.unregister_form();
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _props3 = this.props,
			    children = _props3.children,
			    className = _props3.className,
			    style = _props3.style;
			var rrui__modal = this.context.rrui__modal;


			return _react2.default.createElement(
				'div',
				{
					className: (0, _classnames2.default)('rrui__form__actions', className),
					style: style },
				rrui__modal && _react2.default.createElement(
					_Button2.default,
					{ action: rrui__modal.close_if_not_busy },
					rrui__modal.closeLabel
				),
				children
			);
		}
	}]);
	return Actions;
}(_react.Component);

Form.Actions.contextTypes = {
	rrui__modal: _propTypes2.default.object
};
