// react-responsive-ui/commonjs/Menu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _PageAndMenu = require('./PageAndMenu');

var _PageAndMenu2 = _interopRequireDefault(_PageAndMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// A slideout menu
var Menu = function (_PureComponent) {
	(0, _inherits3.default)(Menu, _PureComponent);

	function Menu() {
		var _ref;

		var _temp, _this, _ret;

		(0, _classCallCheck3.default)(this, Menu);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Menu.__proto__ || (0, _getPrototypeOf2.default)(Menu)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
			show: false
		}, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
	}

	(0, _createClass3.default)(Menu, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var _this2 = this;

			var _context = this.context,
			    router = _context.router,
			    _context$react_respon = _context.react_responsive_ui_menu,
			    register = _context$react_respon.register,
			    toggle = _context$react_respon.toggle;
			var show = this.state.show;

			// if (!slideout)
			// {
			// 	return
			// }

			// If `react-router` is being used

			if (router) {
				// Hide slideout menu on navigation
				this.unlisten_history = router.listen(function (location) {
					if (_this2.state.show) {
						toggle();
					}
				});
			}

			this.unregister = register({
				hide: function hide() {
					return _this2.setState({ show: false });
				},
				toggle: function toggle() {
					return _this2.setState(function (state) {
						return { show: !state.show };
					});
				},
				element: function element() {
					return _reactDom2.default.findDOMNode(_this2.menu);
				}
			});

			// this.calculate_width()
		}

		// componentDidUpdate(previous_props, previous_state)
		// {
		// 	// On menu toggle
		// 	if (this.state.show !== previous_state.show)
		// 	{
		// 		// Requires a corresponding `clearTimeout()`` in `componentWillUnmount()``
		// 		setTimeout(() =>
		// 		{
		// 			this.setState({ page_moved_aside: this.state.show })
		// 		},
		// 		menu_transition_duration)
		//
		// 		this.calculate_width()
		// 	}
		// }

	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			// const { slideout } = this.props

			// if (!slideout)
			// {
			// 	return
			// }

			this.unregister();

			if (this.unlisten_history) {
				this.unlisten_history();
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _this3 = this;

			var _props = this.props,
			    className = _props.className,
			    style = _props.style,
			    children = _props.children;
			var show = this.state.show;


			var markup = _react2.default.createElement(
				'div',
				{
					ref: function ref(_ref2) {
						return _this3.menu = _ref2;
					},
					className: (0, _classnames2.default)('rrui__slideout-menu', {
						'rrui__slideout-menu--shown': show
					}, className),
					style: style },
				children
			);

			return markup;
		}

		// calculate_width()
		// {
		// 	const dom_node = ReactDOM.findDOMNode(this.menu)
		// 	this.props.updateWidth(dom_node.offsetWidth)
		// }

	}]);
	return Menu;
}(_react.PureComponent);

Menu.propTypes = {
	// CSS style object
	style: _propTypes2.default.object,

	// CSS class
	className: _propTypes2.default.string
};
Menu.contextTypes = (0, _extends3.default)({
	router: _propTypes2.default.object }, _PageAndMenu2.default.childContextTypes);
exports.default = Menu;
