// react-responsive-ui/commonjs/Switch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// http://wd.dizaina.net/en/experiments/ios7-style-switch/

// An iOS-style switch
var Switch = function (_PureComponent) {
	(0, _inherits3.default)(Switch, _PureComponent);

	function Switch() {
		(0, _classCallCheck3.default)(this, Switch);

		var _this = (0, _possibleConstructorReturn3.default)(this, (Switch.__proto__ || (0, _getPrototypeOf2.default)(Switch)).call(this));

		_this.state = {};


		_this.toggle = _this.toggle.bind(_this);
		return _this;
	}

	// Client side rendering, javascript is enabled


	(0, _createClass3.default)(Switch, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var fallback = this.props.fallback;


			if (fallback) {
				this.setState({ javascript: true });
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _props = this.props,
			    value = _props.value,
			    fallback = _props.fallback,
			    disabled = _props.disabled,
			    indicateInvalid = _props.indicateInvalid,
			    error = _props.error,
			    style = _props.style,
			    className = _props.className,
			    leftLabel = _props.leftLabel,
			    rightLabel = _props.rightLabel,
			    children = _props.children;


			var markup = _react2.default.createElement(
				'label',
				{
					className: (0, _classnames2.default)('rrui__switch', {
						'rrui__rich': fallback,
						'rrui__input': children,
						'rrui__switch--label': children,
						'rrui__switch--disabled': disabled
					}, className),
					style: style },
				(leftLabel || children) && _react2.default.createElement(
					'div',
					{ className: (0, _classnames2.default)('rrui__switch__label', 'rrui__switch__label--left') },
					leftLabel || children
				),
				_react2.default.createElement(
					'div',
					{ className: 'rrui__switch__switch' },
					_react2.default.createElement('input', {
						type: 'checkbox',
						value: value,
						onChange: this.toggle,
						style: input_style }),
					_react2.default.createElement('div', {
						className: (0, _classnames2.default)('rrui__switch__groove', {
							// CSS selector performance optimization
							'rrui__switch__groove--on': value
						}) }),
					_react2.default.createElement('div', {
						className: (0, _classnames2.default)('rrui__switch__knob', {
							// CSS selector performance optimization
							'rrui__switch__knob--on': value
						}) })
				),
				rightLabel && _react2.default.createElement(
					'div',
					{ className: (0, _classnames2.default)('rrui__switch__label', 'rrui__switch__label--right') },
					rightLabel
				)
			);

			return markup;
		}

		// supports disabled javascript

	}, {
		key: 'render_static',
		value: function render_static() {
			var _props2 = this.props,
			    name = _props2.name,
			    disabled = _props2.disabled,
			    value = _props2.value;


			var markup = _react2.default.createElement(
				'div',
				{ className: 'rrui__rich__fallback' },
				_react2.default.createElement('input', {
					type: 'checkbox',
					name: name,
					disabled: disabled,
					value: value })
			);

			return markup;
		}
	}, {
		key: 'toggle',
		value: function toggle(event) {
			var _props3 = this.props,
			    onChange = _props3.onChange,
			    disabled = _props3.disabled,
			    value = _props3.value;


			if (disabled) {
				return;
			}

			onChange(!value);
		}
	}]);
	return Switch;
}(_react.PureComponent);

Switch.propTypes = {
	// HTML form input `name` attribute
	name: _propTypes2.default.string,

	// Disables the switch
	disabled: _propTypes2.default.bool,

	// Either `true` or `false`
	value: _propTypes2.default.bool.isRequired,

	// Is called when the switch is switched
	onChange: _propTypes2.default.func.isRequired,

	// (exotic use case)
	// Falls back to a plain HTML input
	// when javascript is disabled (e.g. Tor)
	fallback: _propTypes2.default.bool.isRequired,

	// The label to the left of the switch
	leftLabel: _propTypes2.default.string,

	// The label to the right of the switch
	rightLabel: _propTypes2.default.string,

	// A label (analogous to `leftLabel`)
	children: _propTypes2.default.node,

	// CSS style object
	style: _propTypes2.default.object,

	// CSS class
	className: _propTypes2.default.string
};
Switch.defaultProps = {
	value: false,
	fallback: false
};
exports.default = Switch;


var input_style = {
	position: 'absolute',
	opacity: 0,
	margin: 0,
	width: '100%',
	height: '100%'
};
