// @maji/react-prism/lib/Prism.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _prismjs = require('prismjs');

var _prismjs2 = _interopRequireDefault(_prismjs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Prism = function (_PureComponent) {
  _inherits(Prism, _PureComponent);

  function Prism() {
    _classCallCheck(this, Prism);

    return _possibleConstructorReturn(this, (Prism.__proto__ || Object.getPrototypeOf(Prism)).apply(this, arguments));
  }

  _createClass(Prism, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var className = (0, _classnames2.default)(_defineProperty({
        'react-prism': true
      }, 'language-' + this.props.language, !!this.props.language));

      return _react2.default.createElement(
        'pre',
        { ref: function ref(_ref) {
            return _this2.el = _ref;
          }, className: className },
        _react2.default.createElement(
          'code',
          null,
          this.props.children
        )
      );
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.highlightCode();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.highlightCode();
    }
  }, {
    key: 'highlightCode',
    value: function highlightCode() {
      _prismjs2.default.highlightElement(this.el);
    }
  }]);

  return Prism;
}(_react.PureComponent);

Prism.propTypes = {
  children: _propTypes2.default.node.isRequired,
  language: _propTypes2.default.string
};
exports.default = Prism;
module.exports = exports['default'];
