// react-responsive-ui/modules/Menu.js
import _extends from 'babel-runtime/helpers/extends';
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import classNames from 'classnames';

import PageAndMenu from './PageAndMenu';

// A slideout menu

var Menu = function (_PureComponent) {
	_inherits(Menu, _PureComponent);

	function Menu() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, Menu);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Menu.__proto__ || _Object$getPrototypeOf(Menu)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
			show: false
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(Menu, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var _this2 = this;

			var _context = this.context,
			    router = _context.router,
			    _context$react_respon = _context.react_responsive_ui_menu,
			    register = _context$react_respon.register,
			    toggle = _context$react_respon.toggle;
			var show = this.state.show;

			// if (!slideout)
			// {
			// 	return
			// }

			// If `react-router` is being used

			if (router) {
				// Hide slideout menu on navigation
				this.unlisten_history = router.listen(function (location) {
					if (_this2.state.show) {
						toggle();
					}
				});
			}

			this.unregister = register({
				hide: function hide() {
					return _this2.setState({ show: false });
				},
				toggle: function toggle() {
					return _this2.setState(function (state) {
						return { show: !state.show };
					});
				},
				element: function element() {
					return ReactDOM.findDOMNode(_this2.menu);
				}
			});

			// this.calculate_width()
		}

		// componentDidUpdate(previous_props, previous_state)
		// {
		// 	// On menu toggle
		// 	if (this.state.show !== previous_state.show)
		// 	{
		// 		// Requires a corresponding `clearTimeout()`` in `componentWillUnmount()``
		// 		setTimeout(() =>
		// 		{
		// 			this.setState({ page_moved_aside: this.state.show })
		// 		},
		// 		menu_transition_duration)
		//
		// 		this.calculate_width()
		// 	}
		// }

	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			// const { slideout } = this.props

			// if (!slideout)
			// {
			// 	return
			// }

			this.unregister();

			if (this.unlisten_history) {
				this.unlisten_history();
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _this3 = this;

			var _props = this.props,
			    className = _props.className,
			    style = _props.style,
			    children = _props.children;
			var show = this.state.show;


			var markup = React.createElement(
				'div',
				{
					ref: function ref(_ref2) {
						return _this3.menu = _ref2;
					},
					className: classNames('rrui__slideout-menu', {
						'rrui__slideout-menu--shown': show
					}, className),
					style: style },
				children
			);

			return markup;
		}

		// calculate_width()
		// {
		// 	const dom_node = ReactDOM.findDOMNode(this.menu)
		// 	this.props.updateWidth(dom_node.offsetWidth)
		// }

	}]);

	return Menu;
}(PureComponent);

Menu.propTypes = {
	// CSS style object
	style: PropTypes.object,

	// CSS class
	className: PropTypes.string
};
Menu.contextTypes = _extends({
	router: PropTypes.object }, PageAndMenu.childContextTypes);
export default Menu;
